src/dag.rs

```diff
@@ -30,43 +174,414 @@ impl<Types: RepoTypes> IpldDag<Types> {
+        // simple directories, you can find one in the history of ipfs-http.
+
+        // advancing is required here in order for us to determine if this was the last element.
+        // This should be ok as the only way we can continue resolving deeper is the case of Link
+        // being matched, and not the error nor the DagPbData case.
+        let segment = segments.next().unwrap();
+
+        Ok(resolve_local_dagpb(
+            cid,
+            data,
             segment,
-            other
-        )),
+            segments.peek().is_none(),
+            cache,
+        )?)
+    } else {
+        let ipld = match decode_ipld(&cid, &data) {
+            Ok(ipld) => ipld,
+            Err(e) => return Err(RawResolveLocalError::UnsupportedDocument(cid, e.into())),
+        };
+        resolve_local_ipld(cid, ipld, segments)
+    }
+}
+
+/// Resolving through dagpb documents is basically just mapping from [`MaybeResolved`] to the
+/// return value, with the exception that path ending in "Data" is returned as
+/// `ResolvedNode::DagPbData`.
+fn resolve_local_dagpb<'a>(
+    cid: Cid,
+    data: Box<[u8]>,
+    segment: &'a str,
+    is_last: bool,
+    cache: &mut Option<Cache>,
+) -> Result<(LocallyResolved<'a>, usize), RawResolveLocalError> {
+    match resolve(&data, segment, cache) {
+        Ok(MaybeResolved::NeedToLoadMore(lookup)) => {
+            Ok((LocallyResolved::Incomplete(cid, lookup), 0))
+        }
+        Ok(MaybeResolved::Found(dest)) => {
+            Ok((LocallyResolved::Complete(ResolvedNode::Link(cid, dest)), 1))
+        }
+        Ok(MaybeResolved::NotFound) => {
+            if segment == "Data" && is_last {
+                let wrapped = wrap_node_data(data).expect("already deserialized once");
+                return Ok((
+                    LocallyResolved::Complete(ResolvedNode::DagPbData(cid, wrapped)),
+                    1,
+                ));
+            }
+            Err(RawResolveLocalError::NotFound {
+                document: cid,
+                segment_index: 0,
+            })
+        }
+        Err(ipfs_unixfs::ResolveError::UnexpectedType(ut)) if ut.is_file() => {
+            // this might even be correct: files we know are not supported, however not sure if
+            // symlinks are, let alone custom unxifs types should such exist
+            Err(RawResolveLocalError::NotFound {
+                document: cid,
+                segment_index: 0,
+            })
+        }
+        Err(e) => Err(RawResolveLocalError::UnsupportedDocument(cid, e.into())),
+    }
+}
+
+/// Resolves the given path segments locally or inside the given document. Resolving is terminated
+/// upon reaching a link or exhausting the path.
+///
+/// Returns the number of path segments matched -- the iterator might be consumed more than it was
+/// matched.
+///
+/// Note: Tried to initially work with this through Peekable but this would need two peeks.
```

Comment:
I only now remembered that there's `itertools::multipeek` in case this could improve the implementation in the future.

Reply:
not really, it's a terminal branch which needs this, which is why there is no cleanup either. `peek` is enough in most cases though.